### PR TITLE
ACA-51: Implement "Unassign Access within Organisation" Operation

### PR DESCRIPTION
**Pull request by mattnayler**
### JIRA link (if applicable) ###

   [ACA-51](https://tools.hmcts.net/jira/browse/ACA-51) _'Implement "Unassign Access within Organisation" Operation'_

### Change description ###

Create endpoint for Unassign Access within Organisation

> Note: Just a DRAFT PR as this is currently against ACA-59 as it conatins a merge of ACA-59

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203 - Unassign Access Within Organisation.feature [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+      And the request [contains the Case ID of C1 and a missing case ID],
+      And it is submitted to call the [Unassign Access within Organisation] operation of [Manage Case Assignment Microservice],
+
+    Then a negative response is received,
+      And the response has all the details as expected,
+      And a call [by Becky to confirm that Benjamin still has access to the case - C1] will get the expected response as in [S-203.7_Verify_Assignments_In_My_Org].
+
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+# ACA-51 A/C 8
+@S-203.8 @Ignore
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203 - Unassign Access Within Organisation.feature:229]: This AC is obsolete and has been removed.

File: src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203 - Unassign Access Within Organisation.feature [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+      And the request [is made by Benjamin and intends to unassign access to C1 for Becky],
+      And it is submitted to call the [Unassign Access within Organisation] operation of [Manage Case Assignment Microservice],
+
+    Then a negative response is received,
+      And the response has all the details as expected,
+      And a call [to Get Assignments In My Organisation by Becky to verify continued assignment of Becky for C1] will get the expected response as in [S-203.8_Verify_Assignments_In_My_Org].
+
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+# ACA-51 A/C 9
+@S-203.9 @Ignore
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203 - Unassign Access Within Organisation.feature:248]: This AC is obsolete and has been removed.

File: src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203_Prerequisite_Case_Assignment_Seperate_Org_Policies_C1_Benjamin_By_Becky.td.json [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+  "request": {
+    "body": {
+      "case_type_id": "${[scenarioContext][siblingContexts][F-203_Prerequisite_Case_Creation_Seperate_Org_Policies_C1][testData][actualResponse][body][case_type_id]}",
+      "case_id": "${[scenarioContext][siblingContexts][F-203_Prerequisite_Case_Creation_Seperate_Org_Policies_C1][testData][actualResponse][body][id]}",
+      "assignee_id": "${[scenarioContext][testData][users][Benjamin][id]}"
+    }
+  },
+
+	"expectedResponse": {
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203_Prerequisite_Case_Assignment_Seperate_Org_Policies_C1_Benjamin_By_Becky.td.json:26]: This Tab/Space issue should now be resolved

File: src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203_Prerequisite_Case_Assignment_Seperate_Org_Policies_C1_Other_Org2_By_CAA_Org2.td.json [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  "request": {
+    "body": {
+      "case_type_id": "${[scenarioContext][siblingContexts][F-203_Prerequisite_Case_Creation_Seperate_Org_Policies_C1][testData][actualResponse][body][case_type_id]}",
+      "case_id": "${[scenarioContext][siblingContexts][F-203_Prerequisite_Case_Creation_Seperate_Org_Policies_C1][testData][actualResponse][body][id]}",
+      "assignee_id": "${[scenarioContext][testData][users][Other_Org2][id]}"
+    }
+  },
+
+	"expectedResponse": {
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203_Prerequisite_Case_Assignment_Seperate_Org_Policies_C1_Other_Org2_By_CAA_Org2.td.json:26]: This step is obsolete and has been removed.

File: src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203 - Unassign Access Within Organisation.feature [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+
+#-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+# ACA-51 A/C 5
+@S-203.5
+Scenario: Must return an error response if intended unassignee doesn't exist in invoker's organisation
+
+    Given a user [Becky - with a Solicitor role for a particular jurisdiction under an organisation to create, assign and unassign access to a case for another solicitor in their organisation - O1],
+      And a user [Benjamin - with a solicitor role for the same jurisdiction within organisation - O1],
+      And a user [Other_Org2 - with a solicitor role for the same jurisdiction within a different organisation from Becky - O2],
+    #  And a user [CAA_Org2 - with a Pui-CAA role for the same jurisdiction and in organisation - O2],
```
[the remark this answers is not in the document]
Comment (mattnayler) [reply src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203 - Unassign Access Within Organisation.feature:170]: This AC has been updated to remove these steps which cannot be achieved using current wiremock.

File: src/main/java/uk/gov/hmcts/reform/managecase/service/CaseAssignmentService.java [hunk 6]
```diff
@@ -102,6 +99,71 @@ public CaseAssignmentService(PrdRepository prdRepository,
...
+            missingRoleInformation.addAll(dataStoreRepository.getCaseAssignments(caseIds, userIds));
+        }
+
+        List<CaseUserRole> caseUserRolesToUnassign = expandUnassignmentsList(unassignments, missingRoleInformation);
+
+        // NB: no processing required if empty (i.e. no roles found to delete)
+        if (!caseUserRolesToUnassign.isEmpty()) {
+            dataStoreRepository.removeCaseUserRoles(caseUserRolesToUnassign, usersByOrg.getOrganisationIdentifier());
+        }
+    }
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/reform/managecase/service/CaseAssignmentService.java:142]: Better split this into a few methods with nicely self-descriptive names each.

File: src/main/java/uk/gov/hmcts/reform/managecase/api/errorhandling/AuthError.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package uk.gov.hmcts.reform.managecase.api.errorhandling;
+
+/**
+ * Authentication and Authorisation error messages.
+ **/
+public final class AuthError {
+
+    public static final String AUTHENTICATION_TOKEN_INVALID
+        = "Authentication failure due to invalid / expired tokens (IDAM / S2S)";
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/managecase/api/errorhandling/AuthError.java]: Agreed with @mattnayler that this is a standard error response

File: src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/S-203.8.td.json [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+          "assignee_id": "${[scenarioContext][testData][users][Becky][id]}"
+        }
+      ]
+    }
+  },
+
+  "expectedResponse": {
+    "_extends_": "Common_403_Response",
+    "body": {
+      "message" : "The user is neither a case access administrator nor a solicitor with access to the jurisdiction"
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/S-203.8.td.json]: @MSancaktutar What are your thoughts on this - acceptance criteria states

"Must return an error response if the invoker is a solicitor in a different jurisdiction from that of the case."

File: build.gradle [hunk 9]
```diff
@@ -263,6 +263,9 @@ dependencies {
   testImplementation group: 'io.rest-assured', name: 'json-path', version: versions.restAssured
   testImplementation group: 'io.rest-assured', name: 'xml-path', version: versions.restAssured
 
+  // https://mvnrepository.com/artifact/nl.jqno.equalsverifier/equalsverifier
+  testCompile group: 'nl.jqno.equalsverifier', name: 'equalsverifier', version: '3.4.2'
+
```
[the remark this answers is not in the document]
Comment (MSancaktutar) [reply build.gradle:268]: Yes.

**Review by MSancaktutar — APPROVED** [review]
LGTM

**Comment by sullivangareth** [comment]
This ACA-51 branch will need to be merged into `master`

Wait until `ACA-59` has been merged into `master` before merging this PR

File: src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203 - Unassign Access Within Organisation.feature [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+      And a case [by Becky to create a case - C1 with Organisation policies containing R1 and R2] created as in [F-203_Prerequisite_Case_Creation_Seperate_Org_Policies_C1],
```
[the remark this answers is not in the document]
Comment (sullivangareth) [reply src/functionalTest/resources/features/F-203 - Unassign Access Within Organisation/F-203 - Unassign Access Within Organisation.feature:114]: Happy with explanation that R1 and R2 are the roles.

**Pull request merged**